galaxy/worker/tasks/collection.py

```diff
@@ -0,0 +1,143 @@
+# (c) 2012-2019, Ansible by Red Hat
+#
+# This file is part of Ansible Galaxy
+#
+# Ansible Galaxy is free software: you can redistribute it and/or modify
+# it under the terms of the Apache License as published by
+# the Apache Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Ansible Galaxy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# Apache License for more details.
+#
+# You should have received a copy of the Apache License
+# along with Galaxy.  If not, see <http://www.apache.org/licenses/>.
+
+import logging
+import tempfile
+import tarfile
+
+from django.db import transaction
+from django.db.utils import IntegrityError
+from pulpcore.app import models as pulp_models
+
+from galaxy.common import schema
+from galaxy.importer import collection as importer
+from galaxy.importer import exceptions as i_exc
+from galaxy.main import models
+from galaxy.worker import exceptions as exc
+from galaxy.worker import logutils
+
+
+log = logging.getLogger(__name__)
+
+
+def import_collection(artifact_id, repository_id):
+    task = models.CollectionImport.current()
+    log.info('Starting collection import task: {}'.format(task.id))
+
+    artifact = pulp_models.Artifact.objects.get(pk=artifact_id)
+    repository = pulp_models.Repository.objects.get(pk=repository_id)
+
+    filename = schema.CollectionFilename(
+        task.namespace.name, task.name, task.version)
+
+    task_logger = _get_task_logger(task)
+    task_logger.info('Starting import: task_id={}, artifact_id={}'
+                     .format(task.id, artifact_id))
+    try:
+        collection_info = _process_collection(
+            artifact, filename, task_logger)
+        _publish_collection(task, artifact, repository, collection_info)
+    except exc.TaskError as e:
+        artifact.delete()
+        task_logger.error('Import Task "{task_id}" failed: {message}'
+                          .format(task_id=task.id, message=str(e)))
+        raise
+
+    warnings, errors = task.get_message_stats()
+    msg = ('Import completed with {warnings} warnings and {errors} errors'
+           .format(warnings=warnings, errors=errors))
+    task_logger.info(msg)
+
+
+def _get_task_logger(task):
+    logger = logging.getLogger('galaxy.worker.tasks.import_collection')
+    return logutils.ImportTaskAdapter(logger, task=task)
+
+
+def _process_collection(artifact, filename, task_logger):
+    with tempfile.TemporaryDirectory() as extract_dir:
+        with artifact.file.open() as pkg_file, \
+                tarfile.open(fileobj=pkg_file) as pkg_tar:
+            pkg_tar.extractall(extract_dir)
+
+        try:
+            collection_info = importer.import_collection(
```

Comment:
~~The combo of `from galaxy.importer import collection as importer` and `collection_info = import.import_collection()` is a little confusing to me. Especially In combo with the `from galaxy.importer import collection as importer` returns a `CollectionLoader`. 
And the var here is `collection_info` and `collection_info.collection_info.` is used in _log_collection_info.~~

I think the code does the right thing, just a little hard to follow the renames.

Reply:
Nevermind, I missed the `load()` in `import.import_collection().load()`

Reply:
Maybe a rename `importer_obj = importer.import_collection()` and therefore `metadata = importer_obj.collection_info`